### PR TITLE
A few fixes for Pypy

### PR DESCRIPTION
**Pull request by saimn**
Fix test collection (for some reason doctest fails with functions imported from `np.random`], and one fix for a `open` use. But a lot remains, especially with refcounting on which `io.fits` (to know if mmap can be closed) or `Table` relies on.

Ref #7768

### CODE REVIEW

File: astropy/io/ascii/tests/test_write.py [hunk 1]
```diff
@@ -673,7 +673,9 @@ def test_write_quoted_empty_field(fast_writer):
 def test_write_overwrite_ascii(format, fast_writer, tmpdir):
     """Test overwrite argument for various ASCII writers"""
     filename = tmpdir.join("table-tmp.dat").strpath
-    open(filename, 'w').close()
+    with open(filename, 'w'):
+        # create empty file
```
Comment (cdeil) [inline astropy/io/ascii/tests/test_write.py]: Maybe do this?
```
Path(filename).touch()
```
https://docs.python.org/3/library/pathlib.html#pathlib.Path.touch
https://github.com/python/cpython/blob/4ff38870b1de8a3add5357edf125c2866bc42b54/Lib/pathlib.py#L1218

**Comment by cdeil** [comment]
> I'll let @bsipocz milestone this one, as I am unsure how painful it is to backport, or do we even care about PY2 support for Pypy.

Backporting changes to make pypy work would be a huge pain. Suggest to milestone this PR for the next Astropy feature release and do no backporting.

**Comment by bsipocz** [comment]
Even if we don't aim for pypy compatibility and if this is all the changes we'll have in this PR, I would prefer to backport them as a cleanup (so there will be less possible conflicts later on with a bugfix).
Obviously I wouldn't push for any backports where the changes are less trivial, or the straightforward one is not py2 compatible.

**Comment by cdeil** [comment]
@bsipocz - Naively I would have expected that the opposite, that backporting such non-bug PRs creates a ton of work, because then you backport so many PRs. But of course you have more experience with this, do as you think best.

Note that I suggested the use of `pathlib.Path.touch` above. If @saimn chooses to put that, then that's Py3 only.

**Comment by bsipocz** [comment]
@cdeil - Straightforward PRs like this is usually easy to backport, so I prefer to have cleanups and docs changes etc backported, so the diversion between the two branches gets smaller. But, it also regularly happens if I run into unforseen conflicts that the milestone gets changed back from LTS for non essential bugfixes.

Anyway, I agree that this whole thread is moot in case we use `Path.touch` instead.

File: astropy/io/ascii/tests/test_write.py (hunk 1 above)
Comment (pllim) [reply astropy/io/ascii/tests/test_write.py]: @saimn , are you planning to address this or are you leaving it out for PY2 compat?

Comment (saimn) [reply astropy/io/ascii/tests/test_write.py]: @pllim - I prefer the backport option, so keeping the current version.

**Comment by pllim** [comment]
Thanks, everyone!

**Pull request merged**